### DISCUSSION TITLE
How to get touchscreen keyboard?

### DISCUSSION THREAD

**Discussion by RK11111111111** [fayaaz/mixxx-pi-gen#75] [General]
I am starting to get mixxx up and running on the pi.

One thing I haven't figured out is how to get a touchscreen keyboard for text entry.  Especially search.  I know my kids are going to want to listen to "What Does the Fox Say" first thing.  

I have installed either matchbox or onboard.  I can run onboard.  It seems that on a raspian OS there is a menu for editing preferences, where this can be configured.  I don't see how to do that on this particular install, though I can get to the terminal either via ssh or with Ctrl-Enter after exiting mixxx.  It seems that there are .config files that may be edited, however they don't appear to exist. 

Other things that I will be likely thinking about are right click gestures.

**Comment by fayaaz** [comment]
Hello there is actually [wvkbd](https://github.com/jjsullivan5196/wvkbd) installed on the image.

Here's a script I call from i3blocks to get it to show:

/home/pi/.config/i3blocks/scripts/keyboard.sh:
```
if [[ $(pgrep wvkbd-mobintl) ]]  ; then
  kill -s SIGRTMIN $(pgrep wvkbd-mobintl)
else
  nohup /usr/bin/wvkbd-mobintl  -l nav,special,full --landscape-layers nav,special,full &> /dev/null &
fi
```

i3blocks config block (/home/pi/.config/i3blocks/config):
```
[click]
full_text=KeyBoard
command=/bin/bash /home/pi/.config/i3blocks/scripts/keyboard.sh
```

I will add this to the next release of the image. If there's a way to get it to automatically show up when you press the search bar that would be ideal, I think it does that in gnome with certain on screen keyboards, so it may be possible with sway.

**Reply by RK11111111111** [reply]
This worked pretty well.  I think it would be best if mixxx were aware of being used with a touch screen.

Any idea of how to get a right click functionality? This is another thing that appears to be a mixxx issue because it is optimized for fast first clicks.  I'm not sure where all it crops up, but in library mode it appears to be necessary for loading samples to a sampler or sending them to a crate, at least with Pioneer skin.  However, sway and wayland appear to not be helping much.

I saw a patch in a version of mix that allowed mapping a controller button as sort of a shift/ctrl key that would change the click event to a right click event in mixxx.  

Is this something I could use i3 to do [https://major.io/p/tray-icons-in-i3/](url)

Getting close, my kids are going to love it, this is exciting!

**Reply by fayaaz** [reply]
I would look into sway input. I had a quick google and couldn't find the exact answer, but will try hacking around it - lisgd https://github.com/jjsullivan5196/lisgd seems to come up for sway gestures. It might be something custom that each person has to put into their sway config depending on their touchscreen or input device.

**Reply by RK11111111111** [reply]
yeah, that is sort of what I found also, but I wasn't familiar with the sway etc.  I contacted the vender and they don't supply any drivers.  EG they don't send right click events.

One thing I was thinking would be cool, would be to use a gpio button button or encoder.  It could be mounted on the side like an ipad.

**Reply by RK11111111111** [reply]
It seems that libinput has some functionality:

This thread mentions it: https://www.reddit.com/r/swaywm/comments/rrnsxv/how_do_i_disable_the_right_click_when_pressing/

When I read the man page on sway-input there were some relevant looking text:

input <identifier> scroll_method none|two_finger|edge|on_button_down
           Changes the scroll method for the specified input device.

       input <identifier> tap enabled|disabled
           Enables or disables tap for specified input device.

       input <identifier> tap_button_map lrm|lmr
           **Specifies which button mapping to use for tapping. lrm treats 1 finger as left   
           click, 2 fingers as right click, and 3 fingers as middle click.** lmr treats 1     
           finger as left click, 2 fingers as middle click, and 3 fingers as right click.  

I think something needs to be don't about scrolling also, I don't know about your screen but mine is set for dragging, so scrolling the tracks a is a bit onerous.

**Reply by RK11111111111** [reply]
This was useful:

I can get the status of the sway inputs by:

export SWAYSOCK=$(ls /run/user/1000/sway-ipc.* | head -n 1)

swaymsg -t get_inputs

Which gave me the information about the touchscreen:

Input device: WingCool Inc. TouchScreen
  Type: Touch
  Identifier: 10182:2072:WingCool_Inc._TouchScreen
  Product ID: 2072
  Vendor ID: 10182
  Libinput Send Events: enabled

Seems like a matter of using sway input to enable the multifinger gesture mapping.  Might have to test that out.

**Reply by RK11111111111** [reply]
This was also useful, it displayed the events sent by libinput.  My touchscreen does appear to be sending double taps.  They have to be pretty far apart.  It reports them as the first touch (0) and second touch (1)

libinput debug-events 

event2   TOUCH_DOWN              +189.229s     0 (0) 63.12/47.17 (136.08/42.72mm)
 event2   TOUCH_FRAME             +189.229s
 event2   TOUCH_DOWN              +189.791s     1 (1) 75.54/67.03 (162.84/60.71mm)
 event2   TOUCH_FRAME             +189.791s
 event2   TOUCH_UP                +189.888s     1 (1)
 event2   TOUCH_FRAME             +189.888s
 event2   TOUCH_UP                +190.253s     0 (0)
 event2   TOUCH_FRAME             +190.253s

**Comment by RK11111111111** [comment]
And for the mouse:

event3   POINTER_BUTTON          +102.418s     BTN_LEFT (272) pressed, seat count: 1
 event3   POINTER_BUTTON          +102.530s     BTN_LEFT (272) released, seat count: 0
 event3   POINTER_BUTTON          +102.946s     BTN_RIGHT (273) pressed, seat count: 1
 event3   POINTER_BUTTON          +103.058s     BTN_RIGHT (273) released, seat count: 0

**Comment by RK11111111111** [comment]
I think I might be able to add a bindsym to the sway config:

cat /etc/sway/config

It has something like this:

#   input "2:14:SynPS/2_Synaptics_TouchPad" {
#       dwt enabled
#       tap enabled
#       natural_scroll enabled
#       middle_emulation enabled
#   }

**Comment by RK11111111111** [comment]
This is explains why gestures in Touchscreens are not supported, but touch pads are:

https://wayland.freedesktop.org/libinput/doc/latest/gestures.html#gestures-with-enabled-software-buttons

I guess in our instance there are only one window though.

**Comment by RK11111111111** [comment]
looks like I can install lisgd, however it is supposed to run a command?? ydotool...
It would be nice if we could get that up and running.  Maybe map a gesture to show the virtual keyboard like two finger swipe right.

**Comment by RK11111111111** [comment]
There is some code in mixxx to fake right clicks with touch screens:

[https://github.com/mixxxdj/mixxx/blob/031dfded8a19790bf8c7ecd694a9de6031fbb21b/src/widget/wwidget.cpp#L61](url)

Maybe we need to try a different version of mixxx?

**Reply by fayaaz** [reply]
Hmmm this is checking for QT < 6. You could try the mixxx 2.3 provided by Debian and see if you have more luck.

**Reply by RK11111111111** [reply]
Is that an older version?  I think the touchscreen code was only recently added to Mixxx.

**Reply by fayaaz** [reply]
2.3 is the current stable release

**Comment by RK11111111111** [comment]
I wonder if there is anyway to make the context menu just a menu item, or maybe a button on the skin?  For a touch screen, it will still be awkward to navigate that menu.  I would just add the samplers to the library skin, which is what I want, but there are other useful things in that menu like adjusting BPM.

**Reply by RK11111111111** [reply]
I think it should be possible to get the functionality through a skin:

[https://github.com/mixxxdj/mixxx/wiki/Creating-Skins](url)

There is a press and release button mapping:  <ConfigKey>[Recording],toggle_recording</ConfigKey>
    <EmitOnPressAndRelease>true</EmitOnPressAndRelease>
    <ButtonState>LeftButton</ButtonState>

Then in the developer menu, there is a [Library] Grouping with an item "show_track_menu".

I'll have to test that.

**Reply by RK11111111111** [reply]
This worked Mixxx is awesome!

**Comment by RK11111111111** [comment]
The virtual keyboard is working pretty well.  I think it would be better if we could use the gestures library lisgd, to load the keyboard.  The issue I am encountering is there are certain modal windows that take up the entire screen and I can't load the keyboard.  Lisgd didn't work for emulating right click, because it only launches a program, and there isn't a great way to send a mouse event through the messaging system, however I think it could launch a keyboard.

**Reply by fayaaz** [reply]
I will look into lisgd

**Reply by RK11111111111** [reply]
I was able to get lisgd working a little.  It needs some tuning, I used the same script that you used above.  One problem was that it didn't seem to close the keyboard, and kept launching new ones.

Also, I ran into another issue related to the touch screen.  There isn't a good way to delete hot cues.

**Reply by RK11111111111** [reply]
I was customizing my controllers midi when I saw this in the manual.  I think its so you can add buttons to skins that emulate right click.  There are a couple of places this comes up.  For example: hot cues need to be right clicked to delete them, though I think I am just going to use a shift button on the controller.

The [Controls] controls[](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#the-controls-controls)
The [[Controls]](https://manual.mixxx.org/2.3/en/mixxx-control#cap-[Controls]) group contains controls that didn’t fit in any other group.

[Controls]touch_shift[](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[Controls]-touch_shift)
Once enabled, all touch tab events are interpreted as right click. This control has been added to provide touchscreen compatibility in 2.0 and might be replaced by a general modifier solution in the future.

Range
:
binary

Feedback
:
All Widgets

New in version 2.0.0.

**Comment by RK11111111111** [comment]
These context menus are out of control.  Another issue, is that the timeout and close.  On the desktop moving the mouse pointer over the menu keeps the menus open.  However, the touch screen there doesn't have that kind of mouse movement.  Though, I think even on the desktop, it is kind of annoying, because it's pretty easy to get off, the menu area.

I had success binding the one I really wanted to a shifted button combination, so I can just hit the magic shift button on the controller, then the sampler button, and it loads it directly to the sampler and there was already a button to load to different decs.  Pretty nice.   I think those were the worst, because the menu is three layers deep, and there many samplers to try to find.  Also, the deleting a hotcue was easily bound to a shift button.  Hit shift then the hotcue and it is gone.  I don't know why mixxx didn't have these in the first place.

I think gestures for virtual keyboard are still wanted.  The thing is you can't run mixxx in full screen if you need to hit the keyboard button in the tray.

**Comment by fayaaz** [comment]
Hey @RK11111111111 this might be of interest to you:
https://github.com/jjsullivan5196/wvkbd/issues/60

So it is possible to open the keyboard on tapping an input via dbus but not yet with wvkbd. Maybe there's a way to use dbus-monitor to do this? The other option is to try squeekboard.

I have a new image which uses waybar if you want to see if it works better for you.
https://github.com/fayaaz/mixxx-pi-gen/pull/78

**Reply by RK11111111111** [reply]
Interesting, what do you think the best way to manage upgrades are? I'd like to keep up with the new releases.  Unfortunately, it appears I will need to reflash, then reconfigure everything.  Though, it will be easier because I have my skin and controller setup, there are still things that have to be manually applied/reapplied, like these keyboard scripts and audio settings.  Does, your image support touch screen out of the box now?  If not, what are the benefits of upgrading.  I may pass.

**Reply by fayaaz** [reply]
That is a good question - I'd like to make it easier to upgrade from version to version. Currently the easiest way is to reimage an SD card and copy the /home/pi/.mixxx folder to the new install, but if you have extra modifications  these won't go across.

**Reply by RK11111111111** [reply]
Maybe I should fork a branch and or issue a pull request.  I wasn't planning on solving the touchscreen, virtual keyboard, controller and audio issues in a general sense though.   Though it would be pretty easy to add the skin I use and controller.xml&js to the repo.

I think the two basic modifications I have made that I don't want to redo are:

A)Adding virtual keyboard support.
B)Adjusting the audio with alsamixer.

Alsamixer adjustment is less of a priority, but it is a pain to do without a keyboard.  Is there a way to tell Alsa mixer that all audio interfaces should start at 100%?

**Comment by dennisdebel** [comment]
You can use ‘amixer’ (part of Alsa-utils) to programatically set volume,
but you have to iterate/select the right interface etc. Is this an issue
for all soundcards/interfaces in general? Because I do not think it is a
good idea to set every interface to 100% since users will have different
soundcard interfaces with different properties (output levels) or am I
missing something?

**Reply by fayaaz** [reply]
Replying to this has somehow put it in the wrong issue! Should be this one:
https://github.com/fayaaz/mixxx-pi-gen/discussions/77

**Reply by dennisdebel** [reply]
I replied via email, I will remind myself to not do that again!

**Reply by RK11111111111** [reply]
If you read the mixer documentation on gain staging, they say that all digital output levels should be set at 100%.  I just happened to do it trying to figure out why the sound wasn't as loud as I expected. 
 You don't want to try to reamplify the signal later in an analogue circuit, it introduces noise.  Most DJ's have mixers, if they need to balance multiple signals.  I do this for keyboards.  If I didn't, I would still prefer to keep the amplifier turned down as low as possible.  Noise.

**Reply by RK11111111111** [reply]
[https://manual.mixxx.org/2.3/en/chapters/user_interface#interface-gain-knob](url)

**Reply by RK11111111111** [reply]
I guess another option would be to have an icon on the desktop/taskbar to make it easy to adjust.  Kind of like the windows sound.

**Reply by RK11111111111** [reply]
For my wayland synth project I was able to get lisgd to work to close windows in sway.  There is precious little documentation, but basically I added.

exec lisgd ....

to the sway config.  This worked out very well, however when I tried to extend it to opening keyboards by calling the keyboard script, it wouldn't close the keyboard, and just kept opening new ones.  I'm not sure why.  I think it works best sending sway messages, but I haven't figured out ydotool yet.  Needs some permissions and daemons to access the mouse input.
